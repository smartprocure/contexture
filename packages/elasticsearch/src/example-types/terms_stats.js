var _ = require('lodash/fp')
let F = require('futil-js')
var esTwoLevel = require('./esTwoLevelAggregation').result
let { buildRegexQueryForWords } = require('../regex')
let { getField } = require('../fields')
let { metrics, hasValidMetrics } = require('../aggUtils')

module.exports = {
  validContext: context => context.key_field && context.value_field,
  async result(context, search, schema) {
    let field = getField(schema, context.key_field, context.fieldMode)
    let orderPaths = F.arrayToObject(
      _.identity,
      metric => ({
        [metric === 'stats'
          ? `twoLevelAgg.${context.order || 'sum'}`
          : `twoLevelAgg_${metric}.value`]: context.sortDir || 'desc',
      }),
      metrics
    )
    let x = await esTwoLevel(
      _.merge(
        {
          filter_agg:
            context.filter &&
            buildRegexQueryForWords(field, context.caseSensitive)(
              context.filter
            ),
          key_type: 'terms',
          key_data: {
            field,
            size: context.size || 10,
            order:
              hasValidMetrics(context) && _.size(context.include)
                ? orderPaths[
                    _.contains('stats', context.include)
                      ? 'stats'
                      : context.order
                  ]
                : orderPaths['stats'],
          },
          value_type: 'stats',
        },
        context
      ),
      search
    )
    return {
      terms: x.results,
    }
  },
}
