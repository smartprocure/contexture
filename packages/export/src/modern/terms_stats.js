import _ from 'lodash/fp'
import { runWith } from '../utils'

export default ({ service, tree, ...node }) => {
  let { key_field, size = 100 } = node
  let run = node => runWith(service, tree, node)

  let result = {
    getTotalRecords: async () => {
      let result = await run({
        key: 'cardinality',
        type: 'cardinality',
        field: key_field,
        fieldMode: 'autocomplete',
      })
      return _.get('context.value', result)
    },
    async *[Symbol.asyncIterator]() {
      let node = result.node = await run({
        key: 'stats',
        type: 'terms_stats',
        key_field,
        size: size || (await getTotalRecords()),
        ...node
      })
      yield node.context.terms
    }
  }
  return result
}