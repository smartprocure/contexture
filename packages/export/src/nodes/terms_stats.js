import _ from 'lodash/fp'
import { runWith } from '../utils'

export default async ({ service, tree, ...node }) => {
  let { key_field, size = 100 } = node
  let run = node => runWith(service, tree, node)

  let totalRecords = _.get(
    'context.value',
    await run({
      key: 'cardinality',
      type: 'cardinality',
      field: key_field,
      fieldMode: 'autocomplete',
    })
  )

  let terms_stats = {
    getTotalRecords() {
      return totalRecords
    },
    async *[Symbol.asyncIterator]() {
      let node = (terms_stats.node = await run({
        key: 'stats',
        type: 'terms_stats',
        key_field,
        size: size || totalRecords,
        ...node,
      }))
      yield _.get('context.terms', node)
    },
  }
  return terms_stats
}
