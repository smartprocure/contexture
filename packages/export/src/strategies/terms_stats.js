import _ from 'lodash/fp'
import { runWith } from '../utils'

export default async ({ service, tree, ...node }) => {
  let { key_field, size = 100 } = node
  let run = node => runWith(service, tree, node)

  let totalRecords = _.get(
    'context.value',
    await run({
      key: 'cardinality',
      type: 'cardinality',
      field: key_field,
      fieldMode: 'autocomplete',
    })
  )

  let terms_stats = {
    getTotalRecords: () => {
      return totalRecords
    },
    async *[Symbol.asyncIterator]() {
      let node = await run({
        key: 'stats',
        type: 'terms_stats',
        key_field,
        size: size || totalRecords,
        ...node,
      })
      yield node.context.terms
    },
  }
  return terms_stats
}
