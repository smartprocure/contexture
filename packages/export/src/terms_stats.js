import _ from 'lodash/fp'
import { runWith, addIterator} from './utils'

export default ({ service, tree, ...node }) => {
  let { key_field, size = 100 } = node
  let run = node => runWith(service, tree, node)
  let done

  let result = {
    async getTotalRecords() {
      let result = await run({
        key: 'cardinality',
        type: 'cardinality',
        field: key_field,
        fieldMode: 'autocomplete',
      })
      return _.get('context.value', result)
    },
    hasNext: () => !done,
    async getNext() {
      let node = result.node = await run({
        key: 'stats',
        type: 'terms_stats',
        key_field,
        size: size || (await result.getTotalRecords()),
        ...node
      })
      done = true
      return node.context.terms
    }
  }
  return addIterator(result)
}
