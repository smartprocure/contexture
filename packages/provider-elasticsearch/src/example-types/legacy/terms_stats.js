var _ = require('lodash/fp')
var esTwoLevel = require('./esTwoLevelAggregation').result
let { buildRegexQueryForWords } = require('../../utils/regex')
let { getField } = require('../../utils/fields')

let orderField = ({ include, order = 'sum' }) =>
  include
    ? `${_.replace('count', 'value_count', order)}.value`
    : `stats.${order}`
module.exports = {
  validContext: node => node.key_field && node.value_field,
  async result(node, search, schema) {
    let field = getField(schema, node.key_field)
    let x = await esTwoLevel(
      _.merge(
        {
          filter_agg:
            node.filter && buildRegexQueryForWords(field)(node.filter),
          key_type: 'terms',
          key_data: {
            field,
            size: node.size || 10,
            order: { [orderField(node)]: node.sortDir || 'desc' },
          },
          value_type: 'stats',
        },
        node
      ),
      search
    )
    return {
      terms: x.results,
    }
  },
}
