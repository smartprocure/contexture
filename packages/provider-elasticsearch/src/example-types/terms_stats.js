var _ = require('lodash/fp')
var esTwoLevel = require('./esTwoLevelAggregation').result
let { buildRegexQueryForWords } = require('../regex')
let { getField } = require('../fields')

module.exports = {
  validContext: context =>
    context.config.key_field && context.config.value_field,
  async result(context, search, schema) {
    let field = getField(schema, context.config.key_field)
    let x = await esTwoLevel(
      _.merge(
        {
          config: {
            filter_agg: context.config.filter && buildRegexQueryForWords(field, context.config.caseSensitive)(context.config.filter),
            key_type: 'terms',
            key_data: {
              field,
              size: context.config.size || 10,
              order: {
                // Disable nested path checking for now as we don't need it anymore:
                //  key +
                //      may need >inner>inner. when adding additional inner for nested filters per old code
                //      xor paths - if both are nested or both not, the inner one wot be in inner
                //      ((!getNestedPath(context.config.value_field) != !getNestedPath(context.config.key_field)) ? '>inner.' : '.') +
                //  order
                [`twoLevelAgg.${context.config.order || 'sum'}`]:
                  context.config.sortDir || 'desc',
              },
            },
            value_type: 'stats',
          },
        },
        context
      ),
      search
    )
    return {
      terms: x.results,
    }
  },
}
