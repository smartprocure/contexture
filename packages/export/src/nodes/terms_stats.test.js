import _ from 'lodash/fp'
import terms_stats from './terms_stats'

describe('terms_stats', () => {
  let defaultTree = {
    key: 'root',
    children: [
      {
        key: 'terms',
        type: 'terms_stats',
      },
      {
        key: 'results',
        type: 'results',
      },
    ],
  }

  let simpleRecords = ['record1', 'record2', 'record3']

  let getSimpleService = () =>
    jest.fn(tree => {
      _.last(tree.children).context = {
        terms: simpleRecords,
        value: 1337,
      }
      return tree
    })

  let prepareSimpleStrategy = async ({
    strategyParams = {},
    service = getSimpleService(),
  } = {}) => {
    let tree = _.cloneDeep(defaultTree)
    let strategy = await terms_stats({
      service,
      tree,
      ...strategyParams,
    })
    return strategy
  }

  it('retrieves the total records (same as the given size)', async () => {
    let strategy = await prepareSimpleStrategy({
      strategyParams: { size: 1337 },
    })
    expect(strategy.getTotalRecords()).toBe(1337)
  })
  it('retrieves records', async () => {
    let service = getSimpleService()
    let strategy = await prepareSimpleStrategy({
      service,
      strategyParams: {
        size: 0,
        value_field: 'LineItem.TotalPrice',
      },
    })
    let arr = []
    for await (const i of strategy) arr.push(i)
    expect(arr).toEqual([simpleRecords])
    expect(service).toMatchSnapshot()
  })
  it('doesnt throw error when service returns unexpected result', async () => {
    let strategy = await prepareSimpleStrategy({ service: jest.fn(_.identity) })
    let arr = []
    for await (const i of strategy) arr.push(i)
    expect(arr).toEqual([undefined])
  })
})
