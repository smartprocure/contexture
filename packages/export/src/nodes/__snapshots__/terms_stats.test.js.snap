// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`terms_stats retrieves records 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "filterOnly": true,
                "key": "terms",
                "type": "terms_stats",
              },
              Object {
                "filterOnly": true,
                "key": "results",
                "type": "results",
              },
            ],
            "filterOnly": true,
            "key": "root",
          },
          Object {
            "context": Object {
              "terms": Array [
                "record1",
                "record2",
                "record3",
              ],
              "value": 1337,
            },
            "field": undefined,
            "fieldMode": "autocomplete",
            "key": "cardinality",
            "type": "cardinality",
          },
        ],
        "join": "and",
        "key": "root-parent",
        "schema": undefined,
        "type": "group",
      },
    ],
    Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "filterOnly": true,
                "key": "terms",
                "type": "terms_stats",
              },
              Object {
                "filterOnly": true,
                "key": "results",
                "type": "results",
              },
            ],
            "filterOnly": true,
            "key": "root",
          },
          Object {
            "context": Object {
              "terms": Array [
                "record1",
                "record2",
                "record3",
              ],
              "value": 1337,
            },
            "key": "stats",
            "key_field": undefined,
            "size": 1337,
            "type": "terms_stats",
            "value_field": "LineItem.TotalPrice",
          },
        ],
        "join": "and",
        "key": "root-parent",
        "schema": undefined,
        "type": "group",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "filterOnly": true,
                "key": "terms",
                "type": "terms_stats",
              },
              Object {
                "filterOnly": true,
                "key": "results",
                "type": "results",
              },
            ],
            "filterOnly": true,
            "key": "root",
          },
          Object {
            "context": Object {
              "terms": Array [
                "record1",
                "record2",
                "record3",
              ],
              "value": 1337,
            },
            "field": undefined,
            "fieldMode": "autocomplete",
            "key": "cardinality",
            "type": "cardinality",
          },
        ],
        "join": "and",
        "key": "root-parent",
        "schema": undefined,
        "type": "group",
      },
    },
    Object {
      "isThrow": false,
      "value": Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "filterOnly": true,
                "key": "terms",
                "type": "terms_stats",
              },
              Object {
                "filterOnly": true,
                "key": "results",
                "type": "results",
              },
            ],
            "filterOnly": true,
            "key": "root",
          },
          Object {
            "context": Object {
              "terms": Array [
                "record1",
                "record2",
                "record3",
              ],
              "value": 1337,
            },
            "key": "stats",
            "key_field": undefined,
            "size": 1337,
            "type": "terms_stats",
            "value_field": "LineItem.TotalPrice",
          },
        ],
        "join": "and",
        "key": "root-parent",
        "schema": undefined,
        "type": "group",
      },
    },
  ],
}
`;
